File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        rab = head
        tur = head
        while rab and rab.next:
            rab = rab.next.next
            tur = tur.next

            if rab == tur:
                return True
        return False

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int size = 0;
        ListNode* it = head;
        while(it){
            size++;
            it = it->next;
        }

        int idx = 0;
        it = head;
        ListNode* prev = nullptr;
        while(it){
            idx++;
            if(n == size - idx + 1){
                if(!prev){
                    ListNode* newNode = nullptr;
                    delete head;
                    return newNode;
                }
                else{
                    prev->next = it->next;
                    it->next = nullptr;
                    delete it;
                    return head;
                }
            }
            prev = it;
            it = it->next;
        }
    }
};
